snafu-derive/src/lib.rs

```diff
@@ -305,19 +363,41 @@ fn parse_snafu_struct(
 
     let mut transformation = None;
 
+    let mut errors = Vec::new();
+
     for attr in attributes_from_syn(attrs)? {
         match attr {
-            SnafuAttribute::Display(..) => { /* Report this isn't valid here? */ }
-            SnafuAttribute::Visibility(..) => { /* Report this isn't valid here? */ }
+            SnafuAttribute::Display(..) => {
+                errors.push(syn::Error::new(
+                    name.span(),
+                    "'display' attribute is only valid on variants of an enum deriving Snafu; attempted to place on this struct:",
+                ));
+            }
+            SnafuAttribute::Visibility(..) => {
+                errors.push(syn::Error::new(
+                    name.span(),
+                    "'visibility' attribute is only valid on an enum deriving Snafu and its variants; attempted to place on this struct:",
+                ));
+            }
             SnafuAttribute::Source(ss) => {
                 for s in ss {
                     match s {
-                        Source::Flag(..) => { /* Report this isn't valid here? */ }
+                        Source::Flag(..) => {
+                            errors.push(syn::Error::new(
+                                name.span(),
+                                "'source(bool)' is only valid on individual fields; attempted to place on this struct:",
```

Comment:
This one sticks out because it’s the only one that says `(bool)` — why the difference?

Reply:
This one was confusing to me.  I was trying to figure out what `Flag` is, and I think it's when you use `source(false)` to say a field isn't the source even though it's named `source`.  You'd just say `source` if it *is* the source, so `source(true)` shouldn't really be necessary.  That's my understanding of this code path, and I used `source(bool)` as a shorthand for it, since I believe it would only happen if the user did specify a boolean inside `source()`.  Is that right?

Reply:
Ah, yes, I see. Your analysis is correct. I was missing that `source(from(...))` was valid here, but `source(false)` wasn't.

> so `source(true)` shouldn't really be necessary

I did this for consistency of `true` / `false` but also *potentially* allowing the concept of multiple attributes that subsequently override each other. There's also a nascent idea of allowing multiple `source(from())` attributes, for example. I wouldn't worry about either of those future plans, but just so you are aware of the background.